python/hsfs/feature_group.py

```diff
@@ -2750,35 +2756,66 @@ def __init__(
-        self, time_travel_format: Optional[str], online_enabled: bool
+        self,
+        time_travel_format: Optional[str],
+        online_enabled: bool,
+        is_hopsfs: bool,
     ) -> None:
         """Initialize `self._time_travel_format` and `self._stream` for new objects.
 
-        Behavior mirrors the previous inline logic and depends on engine type,
-        provided `time_travel_format`, and `online_enabled`.
+        Extracted into testable helpers to simplify unit testing.
         """
-        if time_travel_format is None:
-            if engine.get_type() == "python":
-                if online_enabled:
-                    self._time_travel_format = "HUDI"
-                    self._stream = True
-                else:
-                    self._time_travel_format = "DELTA"
-            else:
-                self._time_travel_format = "HUDI"
+        self._time_travel_format = FeatureGroup.resolve_time_travel_format(
+            time_travel_format=time_travel_format,
+            online_enabled=online_enabled,
+            is_hopsfs=is_hopsfs,
+        )
+        if engine.get_type() == "python":
+            self._stream = FeatureGroup.resolve_stream_python(
+                time_travel_format=self._time_travel_format,
+                is_hopsfs=is_hopsfs,
+                online_enabled=online_enabled,
+            )
+        return self._time_travel_format, self._stream
 
-        elif time_travel_format == "HUDI":
-            self._time_travel_format = "HUDI"
-            if engine.get_type() == "python":
-                self._stream = True
+    @staticmethod
+    def _is_hopsfs_storage(
+        storage_connector: Optional[sc.StorageConnector],
+    ) -> bool:
+        """Return True if storage is HopsFS."""
+        return storage_connector is None or (
+            storage_connector is not None
+            and storage_connector.type == sc.StorageConnector.HOPSFS
+        )
 
-        elif time_travel_format == "DELTA":
-            self._time_travel_format = "DELTA"
-            if online_enabled and engine.get_type() == "python":
-                self._stream = True
+    @staticmethod
+    def resolve_stream_python(
+        time_travel_format: str,
+        is_hopsfs: bool,
+        online_enabled: bool,
+        *args,
+        **kwargs,
+    ) -> Optional[bool]:
+        if is_hopsfs and time_travel_format == "DELTA" and not online_enabled:
+            return False
+        return True
 
-        else:
-            self._time_travel_format = time_travel_format
+    @staticmethod
+    def resolve_time_travel_format(
+        time_travel_format: Optional[str],
+        online_enabled: bool,
+        is_hopsfs: bool,
+        *args,
+        **kwargs,
+    ) -> str:
+        """Resolve only the time travel format string."""
+        fmt = time_travel_format.upper() if time_travel_format is not None else None
+        if fmt is None:
+            if is_hopsfs and not online_enabled:
```

Comment:
why does online_enabled have any impact here on time travel format?

Reply:
It is just to minimise the breaking changes.